commit cf32565b5768c9bd407a47b06dbbec53a1efd900 (refs/changes/66/5366/6)
Author: Robert Mustacchi <rm@joyent.com>
Date:   2019-01-18T21:02:41+00:00 (9 months ago)
    
    OS-7487 Prefer use of mwait on AMD Zen
    Reviewed by: John Levon <john.levon@joyent.com>
    Reviewed by: Jordan Hendricks <jordan.hendricks@joyent.com>
    Approved by: Patrick Mooney <patrick.mooney@joyent.com>

diff --git a/usr/src/uts/i86pc/os/cpuid.c b/usr/src/uts/i86pc/os/cpuid.c
index ddc09a4951..d0489a0148 100644
--- a/usr/src/uts/i86pc/os/cpuid.c
+++ b/usr/src/uts/i86pc/os/cpuid.c
@@ -2508,13 +2508,14 @@ cpuid_pass1(cpu_t *cpu, uchar_t *featureset)
 
 #if !defined(__xpv)
 		/*
-		 * Do not use MONITOR/MWAIT to halt in the idle loop on any AMD
-		 * processors.  AMD does not intend MWAIT to be used in the cpu
-		 * idle loop on current and future processors.  10h and future
-		 * AMD processors use more power in MWAIT than HLT.
-		 * Pre-family-10h Opterons do not have the MWAIT instruction.
+		 * AMD has not historically used MWAIT in the CPU's idle loop.
+		 * Pre-family-10h Opterons do not have the MWAIT instruction. We
+		 * know for certain that in at least family 17h, per AMD, mwait
+		 * is preferred. Families in-between are less certain.
 		 */
-		idle_cpu_prefer_mwait = 0;
+		if (cpi->cpi_family < 0x17) {
+			idle_cpu_prefer_mwait = 0;
+		}
 #endif
 
 		break;
